### ISSUE TITLE
Support hotplugging ENIs

### ISSUE THREAD

**Issue by mcrute** [mcrute/alpine-ec2-ami#82]
From [tiny-ec2-bootstrap#14](https://github.com/mcrute/tiny-ec2-bootstrap/pull/14#discussion_r475033041). We should support auto-configuration of ENIs in-line with what other distros on AWS support. Right now they just don't get configured at all unless a user does it manually.

**Comment by tomalok** [comment]
Adding/removing the configs for each interface and doing `ifup`/`ifdown` shouldn't be too bad.  The `mdev` hotplug will manage individual config pieces in `/etc/network/interfaces.d` -- if an interface doesn't have a file in there, a default config is written, otherwise the existing one is used.  That should allow further customization if necessary.

Another piece of the puzzle is rules and routes for the interfaces.  I'm not sure at the moment if it belongs here, or in the `udhcpc` hooks (or both), but from an old Gentoo instance that has 4 IPs, 2 per interface, all from the same subnet, I have the following config...
```
~ # cat /etc/conf.d/net
# NOTE: updates to this file are in effect after the next reboot

config_eth0="dhcp
        10.1.3.21/24"
rules_eth0="from 10.1.3.20 table 10
        from 10.1.3.21 table 10"
routes_eth0="default via 10.1.3.1 dev eth0 table 10"
config_eth1="10.1.3.22/24
        10.1.3.23/24"
rules_eth1="from 10.1.3.22 table 11
        from 10.1.3.23 table 11"
routes_eth1="default via 10.1.3.1 dev eth1 table 11"

postup() {
        einfo "Flushing the routing cache"
        ip route flush cache
}

~ # ip route show
default via 10.1.3.1 dev eth0  metric 2 
10.1.3.0/24 dev eth1  proto kernel  scope link  src 10.1.3.22 
10.1.3.0/24 dev eth0  proto kernel  scope link  src 10.1.3.20  metric 2 
127.0.0.0/8 dev lo  scope host 

~ # ip rule show
0:      from all lookup local 
32762:  from 10.1.3.23 lookup 11 
32763:  from 10.1.3.22 lookup 11 
32764:  from 10.1.3.21 lookup 10 
32765:  from 10.1.3.20 lookup 10 
32766:  from all lookup main 
32767:  from all lookup default 
```

If I remember correctly, I needed to do this mapping so bidirectional traffic would return the same way it came in.  It appears that Amazon Linux 2 also maintains association between IPs and interfaces in this way...
```
[root@ip-172-31-8-101 network-scripts]# ip -4 addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc pfifo_fast state UP group default qlen 1000
    inet 172.31.8.101/20 brd 172.31.15.255 scope global dynamic eth0
       valid_lft 3568sec preferred_lft 3568sec
    inet 172.31.6.122/20 brd 172.31.15.255 scope global secondary eth0
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc pfifo_fast state UP group default qlen 1000
    inet 172.31.8.49/20 brd 172.31.15.255 scope global dynamic eth1
       valid_lft 2410sec preferred_lft 2410sec

[root@ip-172-31-8-101 network-scripts]# ip -4 route show
default via 172.31.0.1 dev eth0 
default via 172.31.0.1 dev eth1 metric 10001 
169.254.169.254 dev eth0 
172.31.0.0/20 dev eth0 proto kernel scope link src 172.31.8.101 
172.31.0.0/20 dev eth1 proto kernel scope link src 172.31.8.49 

[root@ip-172-31-8-101 network-scripts]# ip -4 rule show
0:      from all lookup local 
32765:  from 172.31.8.49 lookup 10001 
32766:  from all lookup main 
32767:  from all lookup default 

[root@ip-172-31-8-101 network-scripts]# ip -6 addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 state UP qlen 1000
    inet6 2600:1f14:1cd:1d00:d67b:8fef:6378:70fd/128 scope global dynamic 
       valid_lft 392sec preferred_lft 92sec
    inet6 2600:1f14:1cd:1d00:129e:f357:2b9b:5984/128 scope global dynamic 
       valid_lft 391sec preferred_lft 91sec
    inet6 fe80::820:93ff:fe0f:5fff/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 state UP qlen 1000
    inet6 2600:1f14:1cd:1d00:c2d3:76ab:598b:4b4/128 scope global dynamic 
       valid_lft 391sec preferred_lft 91sec
    inet6 fe80::863:f5ff:fef4:6113/64 scope link 
       valid_lft forever preferred_lft forever

[root@ip-172-31-8-101 network-scripts]# ip -6 route show
unreachable ::/96 dev lo metric 1024 error -113 pref medium
unreachable ::ffff:0.0.0.0/96 dev lo metric 1024 error -113 pref medium
unreachable 2002:a00::/24 dev lo metric 1024 error -113 pref medium
unreachable 2002:7f00::/24 dev lo metric 1024 error -113 pref medium
unreachable 2002:a9fe::/32 dev lo metric 1024 error -113 pref medium
unreachable 2002:ac10::/28 dev lo metric 1024 error -113 pref medium
unreachable 2002:c0a8::/32 dev lo metric 1024 error -113 pref medium
unreachable 2002:e000::/19 dev lo metric 1024 error -113 pref medium
2600:1f14:1cd:1d00:129e:f357:2b9b:5984 dev eth0 proto kernel metric 256 expires 428sec pref medium
2600:1f14:1cd:1d00:c2d3:76ab:598b:4b4 dev eth1 proto kernel metric 256 expires 429sec pref medium
2600:1f14:1cd:1d00:d67b:8fef:6378:70fd dev eth0 proto kernel metric 256 expires 429sec pref medium
2600:1f14:1cd:1d00::/64 dev eth0 proto kernel metric 256 pref medium
2600:1f14:1cd:1d00::/64 dev eth1 proto kernel metric 256 pref medium
unreachable 3ffe:ffff::/32 dev lo metric 1024 error -113 pref medium
fe80::/64 dev eth0 proto kernel metric 256 pref medium
fe80::/64 dev eth1 proto kernel metric 256 pref medium
default via fe80::8a0:a5ff:fe99:2ec dev eth0 proto ra metric 1024 expires 1790sec hoplimit 64 pref medium
default via fe80::8a0:a5ff:fe99:2ec dev eth1 proto ra metric 1024 expires 1795sec hoplimit 64 pref medium

[root@ip-172-31-8-101 network-scripts]# ip -6 rule show
0:      from all lookup local 
32765:  from 2600:1f14:1cd:1d00:c2d3:76ab:598b:4b4 lookup 10001 
32766:  from all lookup main 
```
...seems to be a lot of extra IPv6 noise on `lo` for some reason?

**Comment by tomalok** [comment]
It appears that IP4s and IP6s on `eth0` don't need rules, because it's always going to be the default.

**Comment by tomalok** [comment]
I've made significant progress over the long weekend, but I'm currently somewhat blocked on a `busybox` `iproute2` bug that doesn't show IPv6 routes associated with a custom table...
```
ip -6 route show table 1001
```
...returns nothing after we've created one.  I can submit a patch for Alpine's `busybox` package (I found a one-line fix that never got turned into a patch) -- but that would keep this from working on anything other than edge (until 3.13 happens).  I may still find a way around this that won't be too bad -- the `uchdpc` script is only called when a primary IP is bound or renewed.

On the `mdev` side of things, there's just some cleanup of `ip rule`s and `/run/ifstate` to do when an ENI is removed.

And some more thorough testing.
